packages/eslint-plugin/src/rules/typedef.ts

```diff
@@ -0,0 +1,242 @@
+import { TSESTree, AST_NODE_TYPES } from '@typescript-eslint/typescript-estree';
+import * as ts from 'typescript';
+import * as util from '../util';
+
+interface Options {
+  arrayDestructuring?: boolean;
+  arrowCallSignature?: boolean;
+  arrowParameter?: boolean;
+  callSignature?: boolean;
+  memberVariableDeclaration?: boolean;
+  objectDestructuring?: boolean;
+  parameter?: boolean;
+  propertyDeclaration?: boolean;
+  variableDeclaration?: boolean;
+}
+type Option = keyof Options;
+
+const OPTION_ARRAY_DESTRUCTURING: Option = 'arrayDestructuring';
```

Comment:
would this be better as an enum?
```ts
const enum OptionKeys {
    arrayDestructuring = 'arrayDestructuring',
    arrowCallSignature = 'arrowCallSignature',
    // ....
}

type Options = {
    [k in OptionKeys]?: boolean;
};
```
[repl](https://www.typescriptlang.org/play/#src=const%20enum%20OptionKeys%20%7B%0A%20%20%20%20arrayDestructuring%20%3D%20'arrayDestructuring'%2C%0A%20%20%20%20arrowCallSignature%20%3D%20'arrowCallSignature'%2C%0A%20%20%20%20%2F%2F%20....%0A%7D%0A%0Atype%20Options%20%3D%20%7B%0A%20%20%20%20%5Bk%20in%20OptionKeys%5D%3F%3A%20boolean%3B%0A%7D%3B%0Aconst%20x%3A%20Options%20%3D%20%7B%0A%20%20%20%20%5BOptionKeys.arrayDestructuring%5D%3A%20true%2C%0A%20%20%20%20%5BOptionKeys.arrowCallSignature%5D%3A%20true%0A%7D%0A)

Reply:
Haha probably - I went off the old TSLint file's contents, which were written this way before string enum values.